README.md
# Ractive AutoComplete

An autocomplete input box built on [Ractive](https://github.com/ractivejs/ractive)

## Where to get it?

Racive AutoComplete is available as a [giblet](https://github.com/evs-chris/gobble-giblet), a [component](https://github.com/componentjs/component), and a pre-assembled UMD module. Each flavor does not declare an explicit dependency on Ractive, but it is expected to be available as a global.

All of the pre-built files live in tags on the build branch.

### Development

Ractive AutoComplete uses [gobble](https://github.com/gobblejs/gobble) as its build tool, which makes it easy to build and play around with. The default mode in the gobble file is `development`, which will automatically pull in the edge version of Ractive and make it available along with the sandbox. There is an example file provided along with the source, which you can access by running gobble and pointing you browser at http://localhost:4567/sandbox/example.html.

## Usage

To use AutoComplete, add component reference to your template.

### Attributes

* `completeWith` - method or parameter - a function that receives the current text and returns either an array of completions or a `Promise` for an array of completions.

* `displayMember` - parameter - Either:
  1. The name of the member of a completion object to use as text.
  2. A function that receives a completion value and returns a string.

* `value` - parameter - the current value for the input.

### Events

* `comlpete` - fires when a completion is selected. Parameters:
  1. `text` - the text of the selected completion
  2. `index` - the index of the selected completion
  3. `value` - the selected completion

### Client

A half second after text is entered, the completion function is called for completions. Once the completions are provided, the completion popup is shown for selection. The up and down arrow keys navigate among the available completions. Enter will select one, and escape will close the popup. Typing further will cause the completion function to be called again to refine the completions.

If the input is focused, pressing the down arrow will cause the completion popup to be shown with the last list of completions.


CHANGELOG.md
## 0.4.0

* The component now uses an isolated scope.
* Fixes placeholder handling
* Removes Ractive deprecation warnings
* The package manifest is now available on the build branch.

## 0.3.1

* Removes leftover debug logging statement

## 0.3.0

* Now uses laziness built into Ractive instead of DelayedInput
* There is now a value binding available to use instead of having to listen to `complete` events for changes.
* `displayMember` may now be a function, and there is no `valueMember` any more.
* `completeWith` may now be bound as a component parameter.

## 0.2.0

* Adds support for having different value and display members.

## 0.1.0

* Adds support for promised completions. When the complete function returns a promise, completions will be set from its resolution.
* Adds a complete event that fires when a selection is indicated: on blur, click, and enter.
* The default style will not limit the width of completions, but care must still be taken in fixed with containers with overflow control.

## 0.0.2

* Exports DelayedInput along with AutoComplete

## 0.0.1

Initial version


gobblefile.js
var gobble = require('gobble');
var sander = require('sander');
var path = require('path');

var parts = [], dev = gobble.env() !== 'production';

var transpilerOptions = {
  blacklist: ['es6.modules', 'useStrict']
};

// in dev mode, make libraries and the sandbox available
if (dev) {
  parts.push(gobble('sandbox').moveTo('sandbox'));
  parts.push(gobble('giblets').transform('giblets').moveTo('lib'));
}

var source = gobble('src');

var es5source = gobble(source).transform('babel', transpilerOptions);

// build the main UMD module
var moduled = gobble(es5source)
  .transform('esperanto-bundle', {
    entry: 'index.js',
    dest: 'ractive-autocomplete.js',
    type: 'umd',
    name: 'RactiveAutocomplete',
    strict: true
  });

// for production, build the component too
if (!dev) {
  parts.push(gobble(es5source).transform('esperanto-bundle', {
    entry: 'index.js',
    dest: 'component/index.js',
    type: 'cjs',
    strict: true
  }));
}

parts.push(moduled.
  transform('relative-sourcemaps', { prefix: 'ractive-autocomplete' })
);

// copy the assets
parts.push(gobble('assets').moveTo('assets'));

module.exports = gobble(parts);


component/index.js
'use strict';

/* global Ractive */

var template = '<div class="ractive-autocomplete"><input value="{{text}}" on-blur="blurred()" on-dblclick="popup()" on-keydown="keydown" {{#if .placeholder}}placeholder="{{.placeholder}}"{{/if}} /><div on-click="clicked">\n{{#popup}}\n<ul>{{#completions:i}}<li {{#isCurrent(i)}}class="current"{{/}}>{{display(.)}}</li>{{/}}</ul>\n{{/}}</div></div>';

function isCurrent(idx) {
  return this.get('current') === idx;
}
function index__display(member) {
  return this.display(member);
}

var index = Ractive.extend({
  template: template,
  lazy: 500,
  isolated: true,
  oninit: function oninit() {
    var _this = this;

    this.observe('text', function (cur, prev) {
      // only do completion lookup if this is not from a `select`
      if (cur !== _this.get('currentText')) {
        // reset current index
        _this.set('current', -1);

        // get new completions
        var fn = _this.get('complete');
        if (fn && typeof fn === 'function') {
          var res = fn(cur);
          if (Object.prototype.toString.call(res) === '[object Array]') {
            _this.set('completions', res);
            _this.popup();
          } else if (res.then && typeof res.then === 'function') res.then(function (arr) {
            _this.set('completions', arr);
            _this.popup();
          });
        }
      }
    }, { init: false });
  },
  onrender: function onrender() {
    var _this2 = this;

    var cpl, val;

    this.on('clicked', function (e) {
      var i,
          child,
          ul = _this2.find('ul'),
          found = -1;
      for (i = 0; i < ul.children.length && (child = ul.children[i]); i++) {
        if (child === e.original.target) {
          found = i;
          break;
        }
      }
      _this2.select(found);
    });

    this.on('keydown', function (e) {
      var k = e.original.keyCode;
      if (k === 27) {
        _this2.closePopup();
      } else if (k === 38 || k === 40) {
        e.original.preventDefault();
        if (!_this2.get('popup')) return _this2.popup();
        var c = _this2.get('current');
        _this2.select(k === 38 ? c - 1 : c + 1, false);
      } else if (k === 13) {
        e.original.preventDefault();
        _this2.closePopup();
        _this2.select(_this2.get('current'));
      }
    });

    if ((cpl = this.get('completeWith')) && typeof cpl === 'function') {
      this.completeWith(cpl);
    }

    // if there are no completions but there is a value, use it and select it
    if (!!(cpl = this.get('completions')) && (val = this.get('value'))) {
      this.set('completions', [val]);
      this.select(0, false);
    }
  },
  data: function data() {
    return {
      isCurrent: isCurrent,
      display: index__display,
      completions: [],
      current: -2,
      popup: false,
      currentText: '',
      currentVal: '',
      text: ''
    };
  },
  blurred: function blurred() {
    var _this3 = this;

    if (this.blurTime) {
      clearTimeout(this.blurTime);
    }

    this.blurTime = setTimeout(function () {
      return _this3.closePopup();
    }, 200);
  },
  closePopup: function closePopup() {
    this.set('popup', false);
  },
  popup: function popup() {
    this.set('popup', true);
  },
  display: function display(member) {
    var name = this.get('displayMember');
    if (typeof name === 'string') {
      return member ? member[name] : '';
    } else if (typeof name === 'function') {
      return member ? name(member) : '';
    } else {
      return member ? member : '';
    }
  },
  select: function select(idx) {
    var fin = arguments[1] === undefined ? true : arguments[1];

    var val = this.get('completions') || [];
    var text = '';

    if (idx === -1) idx = val.length - 1;else if (idx === val.length) idx = 0;

    if (val) val = val[idx];
    text = this.display(val);

    // update current selections
    this.set({
      currentText: text,
      currentVal: val
    });

    // update input and highlight
    this.set({
      text: text,
      current: idx
    });

    if (fin) {
      this.closePopup();
      this.set('value', val);
      this.fire('complete', text, idx, val);
    }
  },
  completeWith: function completeWith(fn) {
    this.set('complete', fn);
    this.updateModel('text');
  }
});

exports['default'] = index;
//# sourceMappingURL=index.js.map


ractive-autocomplete.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
  typeof define === 'function' && define.amd ? define(['exports'], factory) :
  factory((global.RactiveAutocomplete = {}))
}(this, function (exports) { 'use strict';

  /* global Ractive */

  var template = '<div class="ractive-autocomplete"><input value="{{text}}" on-blur="blurred()" on-dblclick="popup()" on-keydown="keydown" {{#if .placeholder}}placeholder="{{.placeholder}}"{{/if}} /><div on-click="clicked">\n{{#popup}}\n<ul>{{#completions:i}}<li {{#isCurrent(i)}}class="current"{{/}}>{{display(.)}}</li>{{/}}</ul>\n{{/}}</div></div>';

  function isCurrent(idx) {
    return this.get('current') === idx;
  }
  function index__display(member) {
    return this.display(member);
  }

  var index = Ractive.extend({
    template: template,
    lazy: 500,
    isolated: true,
    oninit: function oninit() {
      var _this = this;

      this.observe('text', function (cur, prev) {
        // only do completion lookup if this is not from a `select`
        if (cur !== _this.get('currentText')) {
          // reset current index
          _this.set('current', -1);

          // get new completions
          var fn = _this.get('complete');
          if (fn && typeof fn === 'function') {
            var res = fn(cur);
            if (Object.prototype.toString.call(res) === '[object Array]') {
              _this.set('completions', res);
              _this.popup();
            } else if (res.then && typeof res.then === 'function') res.then(function (arr) {
              _this.set('completions', arr);
              _this.popup();
            });
          }
        }
      }, { init: false });
    },
    onrender: function onrender() {
      var _this2 = this;

      var cpl, val;

      this.on('clicked', function (e) {
        var i,
            child,
            ul = _this2.find('ul'),
            found = -1;
        for (i = 0; i < ul.children.length && (child = ul.children[i]); i++) {
          if (child === e.original.target) {
            found = i;
            break;
          }
        }
        _this2.select(found);
      });

      this.on('keydown', function (e) {
        var k = e.original.keyCode;
        if (k === 27) {
          _this2.closePopup();
        } else if (k === 38 || k === 40) {
          e.original.preventDefault();
          if (!_this2.get('popup')) return _this2.popup();
          var c = _this2.get('current');
          _this2.select(k === 38 ? c - 1 : c + 1, false);
        } else if (k === 13) {
          e.original.preventDefault();
          _this2.closePopup();
          _this2.select(_this2.get('current'));
        }
      });

      if ((cpl = this.get('completeWith')) && typeof cpl === 'function') {
        this.completeWith(cpl);
      }

      // if there are no completions but there is a value, use it and select it
      if (!!(cpl = this.get('completions')) && (val = this.get('value'))) {
        this.set('completions', [val]);
        this.select(0, false);
      }
    },
    data: function data() {
      return {
        isCurrent: isCurrent,
        display: index__display,
        completions: [],
        current: -2,
        popup: false,
        currentText: '',
        currentVal: '',
        text: ''
      };
    },
    blurred: function blurred() {
      var _this3 = this;

      if (this.blurTime) {
        clearTimeout(this.blurTime);
      }

      this.blurTime = setTimeout(function () {
        return _this3.closePopup();
      }, 200);
    },
    closePopup: function closePopup() {
      this.set('popup', false);
    },
    popup: function popup() {
      this.set('popup', true);
    },
    display: function display(member) {
      var name = this.get('displayMember');
      if (typeof name === 'string') {
        return member ? member[name] : '';
      } else if (typeof name === 'function') {
        return member ? name(member) : '';
      } else {
        return member ? member : '';
      }
    },
    select: function select(idx) {
      var fin = arguments[1] === undefined ? true : arguments[1];

      var val = this.get('completions') || [];
      var text = '';

      if (idx === -1) idx = val.length - 1;else if (idx === val.length) idx = 0;

      if (val) val = val[idx];
      text = this.display(val);

      // update current selections
      this.set({
        currentText: text,
        currentVal: val
      });

      // update input and highlight
      this.set({
        text: text,
        current: idx
      });

      if (fin) {
        this.closePopup();
        this.set('value', val);
        this.fire('complete', text, idx, val);
      }
    },
    completeWith: function completeWith(fn) {
      this.set('complete', fn);
      this.updateModel('text');
    }
  });

  exports['default'] = index;

}));
//# sourceMappingURL=ractive-autocomplete.js.map
